### ISSUE TITLE
Poor performance

### ISSUE THREAD

**Issue by RodrigoMedeirosrRS** [tobspr/RenderPipeline#77]
Hello people!

At the begin of this year i made a complete upgrade on my devolpment hardware, now my specs are:

Processor: AMD APU 7890K
Ram: 16Gb
Video: R7 250 - 2GB
OS: Windows 10 64bits
SSD 120gb

With this config i can running the all Unreal 4 demos in Cinematic Quality but, i can't run the RP into HD resolutions (1280 x 720), when i try it i se many many lags :\

Any suggestion to improve my performance?

**Comment by tobspr** [comment]
Hi,

can you run pstats and see which parts of the pipeline are running slow?

**Comment by RodrigoMedeirosrRS** [comment]
Spstatus + Sys Monitor

1280 x 720
![1280](https://cloud.githubusercontent.com/assets/7956669/24831893/c9b1fbb4-1c79-11e7-9e15-859b818931ed.png)

800 x 600
![800](https://cloud.githubusercontent.com/assets/7956669/24831912/4f837218-1c7a-11e7-87cc-4f896397f605.png)

In both cases, i have many, many lags when i move the camera, but, the frame rate never drops.

**Comment by rdb** [comment]
It would be very helpful to see the pstats charts.  If you have the pstats process running, press 1 in the pipeline to enable it.

**Comment by RodrigoMedeirosrRS** [comment]
Sorry, i confuse the pstats and stats gui.

Here is the PStats:

1280 x 720
![ss1](https://cloud.githubusercontent.com/assets/7956669/25026615/9c57eea0-207d-11e7-9d9f-1a36e26a7822.png)

800 x 600
![ss](https://cloud.githubusercontent.com/assets/7956669/25026657/dc246f7c-207d-11e7-8292-fd7b3b89e435.png)

**Comment by tobspr** [comment]
Could you open the gpu graph and check which parts take  the most time?

**Comment by rdb** [comment]
Also, keep in mind you can double-click the labels on the left to get a more detailed subdivision of that category of the graph.  For example, clicking the "Draw" label will open up a new chart showing how the draw time is divided up.

**Comment by RodrigoMedeirosrRS** [comment]
Note: My pc use's a APU technology - https://www.youtube.com/watch?v=ER_u3yLccBU

So i have made a print for every graph option i will post the most critical first:

**1280 x 720**
_* AMD(TM)R7 Graphics:_
![1280x720 - amd tm r7 graphics](https://cloud.githubusercontent.com/assets/7956669/25258171/e22250ec-2613-11e7-8ba7-29e68ea6c1f6.jpg)

_* Draw_
![1280x720 - draw](https://cloud.githubusercontent.com/assets/7956669/25258194/12fa580e-2614-11e7-8aef-5f8a8e9c672d.jpg)

_* Draw -> Primitive_
![1280x720 - draw - primitive](https://cloud.githubusercontent.com/assets/7956669/25258216/3b8a1714-2614-11e7-83cc-a543640f0187.jpg)

_* App_
![1280x720 - app](https://cloud.githubusercontent.com/assets/7956669/25258228/48e6b41c-2614-11e7-8f24-a537137b4ff1.jpg)

_* Cull_
![1280x720 - cull](https://cloud.githubusercontent.com/assets/7956669/25258238/574cb0b0-2614-11e7-9dce-1cef7dd7eabf.jpg)

_* Time_
![1280x720 - time](https://cloud.githubusercontent.com/assets/7956669/25258247/64216164-2614-11e7-8ad7-12e121bb6ec7.jpg)

_* Wait_
![1280x720 - wait](https://cloud.githubusercontent.com/assets/7956669/25258266/7b1d7fb0-2614-11e7-89ba-acd98473d6c3.jpg)

**800 x 600**
_* AMD(TM)R7 Graphics:_
![800x600 - amd tm r7 graphics](https://cloud.githubusercontent.com/assets/7956669/25258285/976ed0e2-2614-11e7-9058-365af63774a8.jpg)

_* Draw_
![800x600 - draw](https://cloud.githubusercontent.com/assets/7956669/25258306/cb06b320-2614-11e7-8c8f-f5eb94945001.jpg)

_* Draw -> Primitive_
![800x600 - draw - primitive](https://cloud.githubusercontent.com/assets/7956669/25258295/b3d4e08c-2614-11e7-87f0-22b4150ee139.jpg)

_* App_
![800x600 - app](https://cloud.githubusercontent.com/assets/7956669/25258316/d225bd36-2614-11e7-8ee0-de083e070b80.jpg)

_* Cull_
![800x600 - cull](https://cloud.githubusercontent.com/assets/7956669/25258318/d6f50e2a-2614-11e7-9014-7fe4d6d63c43.jpg)

_* Time_
![800x600 - time](https://cloud.githubusercontent.com/assets/7956669/25258324/dbcb5df0-2614-11e7-8857-e3aa5cf065c1.jpg)

_* Wait_
![800x600 - wait](https://cloud.githubusercontent.com/assets/7956669/25258328/e216dfc2-2614-11e7-8674-a98606d6896f.jpg)

**Comment by tobspr** [comment]
Thats weird, it looks like everything is too slow. Are you sure you are not running it on an integrated gpu or so?

**Comment by rdb** [comment]
Yes, an APU has integrated graphics, so make sure that it has been switched to use the external graphics card.

Are these the CPU charts you're showing or the GPU charts?  (The GPU charts are available under the "AMD Radeon… > Frame" menu.

Which version of Panda do you have?

Is there any improvement from adding 'gl-version 3 2' in the pipeline Config.prc?

**Comment by RodrigoMedeirosrRS** [comment]
Yes, i am using the internal Graphics, my internal graphic is a Readeon R7 250 with 2gb and 512 stream processors, my internal Graphics are running the Unreal 4 into cinematic quality and CryEngine 5 in best no 4k quality. The external card works only into the Crossfire, but the AMD crossfire crashe's all OpenGL applications.

**Comment by funcelot** [comment]
I have 1080 ti, and funny part that in small window, it is 400fps+ but in 4k, any scene of almost any complex is no more than 40 fps, seems that python doing some weird staff with paints windows background, cause it is bitbltm the more sureface are, the more ms is spend on purpouseless repaint of transparent background, anyway this is a window and borderless window preoblem on any of the samples. It looks like this is a problem with Pandas3d itself, can you take a look and fix it? @tobspr

**Comment by rdb** [comment]
@funcelot that has nothing to do with Python or Panda3D, but it's because the RenderPipeline is using deferred rendering.  That means most of the shaders run in postprocessing, once per fragment, and so any doubling in resolution will result in a quadrupling of the render time of the entire postprocessing pipeline, which adds up quickly.

4K has a *lot* of pixels to render.  Most games that use deferred rendering do not run well in 4K, and typically render downscaled and then upscale in a postprocessing pass.

**Comment by funcelot** [comment]
> @funcelot that has nothing to do with Python or Panda3D, but it's because the RenderPipeline is using deferred rendering. That means most of the shaders run in postprocessing, once per fragment, and so any doubling in resolution will result in a quadrupling of the render time of the entire postprocessing pipeline, which adds up quickly.
> 
> 4K has a _lot_ of pixels to render. Most games that use deferred rendering do not run well in 4K, and typically render downscaled and then upscale in a postprocessing pass.

Is that the same technique i seen in Cyberpunk2077 wher image seems crispy and make somewhat more detaild in one sec or so? Got your point, thanx

**Comment by funcelot** [comment]
> @funcelot that has nothing to do with Python or Panda3D, but it's because the RenderPipeline is using deferred rendering. That means most of the shaders run in postprocessing, once per fragment, and so any doubling in resolution will result in a quadrupling of the render time of the entire postprocessing pipeline, which adds up quickly.
> 
> 4K has a _lot_ of pixels to render. Most games that use deferred rendering do not run well in 4K, and typically render downscaled and then upscale in a postprocessing pass.

I saw down scaling & posptrocess upscaling in Horizon Zero Dawn, where i play in 4K, but can postprocess in 1080p, using 50% or 25% scale

**Comment by funcelot** [comment]
Can you provide upscaling / downscaling example using you pipline renderer @tobspr ? That would be really aswesome to see 120 fps on 4K, indeed :)

**Comment by tobspr** [comment]
There are multiple ways of doing that, IIRC the pipeline supports it natively though, you can set `resolution_scale` in the config:
https://github.com/tobspr/RenderPipeline/blob/master/config/pipeline.yaml#L26

To e.g. 0.75, the pipeline will then render at 75% resolution and scale up
